### ISSUE TITLE
nix-env -q only lists home-manager-path

### ISSUE THREAD

**Issue by pmiddend** [rycee/home-manager#266]
While going through the Nix Pills series, I noticed that `nix-env -q` only lists `home-manager-path` on my system. Other nix commands are "broken", too. What's wrong here?

**Comment by Infinisil** [comment]
What do you mean by "other commands are broken too"? What you just described regarding `nix-env` is the intended behaviour: home-manager creates a single package combining all your home-manager packages, then uses `nix-env` to install it.

**Comment by pmiddend** [comment]
For example, to find a package, I used to be able to call `nix-env -qaP | grep ...`. This doesn't work anymore.

**Comment by gmarmstrong** [comment]
`nix-env -q` should only list `home-manager-path`, but `nix-env -qaP` should still output a list of packages. Have you fetched the latest packages? Try running `home-manager -n build --update` first.

**Comment by pmiddend** [comment]
I did both now, same result.

**Comment by rycee** [comment]
@pmiddend Hmm, it should work:

```console
$ nix-env -qaP | grep wireshark
nixos.tshark                                                           wireshark-cli-2.4.5
nixos.wireshark-gtk                                                    wireshark-gtk-2.4.5
nixos.wireshark                                                        wireshark-qt-2.4.5
$ nix-env -qaP | wc -l
15078
```
Maybe something is odd about your `$NIX_PATH`? Or maybe your channel is not actually downloaded. Would you mind making a comment with the output of the two commands
```console
$ echo $NIX_PATH
```
and
```console
$ nix-env -v -qaP 2>&1 | head -5
```
?

Personally I don't really use channels anymore so I would run this command as `nix-env -f '<nixpkgs>' -qaP` to use my local Nixpkgs checkout. Perhaps try that command?

**Comment by pmiddend** [comment]
Sure! `echo $NIX_PATH` outputs:

    /home/philipp/.nix-defexpr/channels:nixpkgs=/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs:nixos-config=/etc/nixos/configuration.nix:/nix/var/nix/profiles/per-user/root/channels

And the other command just:

    evaluating file '/nix/store/vdvla43ppjf6gfsi8nx1zmac7rwq72yd-nix-2.0.2/share/nix/corepkgs/derivation.nix'

Out of curiosity, can you elaborate on "you don't use channels anymore"?

**Comment by rycee** [comment]
I'm guessing you don't have a channel downloaded. You could try deleting (or moving out of the way) the `/home/philipp/.nix-defexpr` directory. Then run `nix-channel --update` as root.

When you log in again Nix should automatically recreate the `/home/philipp/.nix-defexpr` directory and hopefully `nix-env -qaP` will work after that.

By me not using channels any more I particularly mean that I no longer use the `nix-channel` command. I do still use channels in the sense that I follow the `nixos-18.03` branch of [nixpkgs-channels](https://github.com/NixOS/nixpkgs-channels). This branch is updated when the release is updated by the hydra builder so I'll pretty much always get packages from the official cache.

In my NixOS configuration I then have something like

```
nix.nixPath = [
  "nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/nixos-18.03.tar.gz"
  "nixos-config=/etc/nixos/configuration.nix"
];
```

and when I `nixos-rebuild switch` it'll download the channel. Basically it's the same as if you have
```
programs.home-manager.path = https://github.com/rycee/home-manager/archive/release-18.03.tar.gz;
```
in your Home Manager configuration.

The above is a bit of a simplification, I actually use my own fork of `nixpkgs-channels` that includes some custom modifications. And on some computers I set `"nixpkgs=/path/to/checkout/of/my/nixpkgsfork"` instead of fetching from github.

**Comment by rycee** [comment]
@pmiddend Yes, if you set it to the official nixos-unstable branch tarball then doing `nixos-rebuild switch` might run you into trouble :-) But if something goes wrong you could typically change the tarball URL to point to a specific revision that you know works. But still, yeah a slight risk.

That's part of the reason why I use a fork of this repo where I have a branch that I keep up to date with the nixos-18.03 channel branch. I run my main laptop against a local clone of this repo and that computer will be updated to a new channel release first. When I'm happy with it I just push the updated branch to my github repo from which other systems of mine will get it.

Not sure if the above description is particularly clear. But specifically I have a branch [rycee-nixos-18.03](https://github.com/rycee/nixpkgs/tree/rycee-nixos-18.03) in my Nixpkgs fork. Occasionally I rebase this branch onto the latest nixos-18.03 branch from [nixpkgs-channels](https://github.com/NixOS/nixpkgs-channels).

**Comment by rollf** [comment]
For me, ``nix-env -qaP`` works fine. But when I run ``nix-env -q`` I do not get any of the packages I installed through home-manager. I am running openSUSE. ``nix-env -q`` only lists packages I installed by hand via ``nix-env`` itself.

This is problematic because I cannot query the nix store. For example, I have installed ``htop`` via home-manager and ``tree`` with ``nix-env``:


```
$ nix-env --installed --no-name --out-path --query tree > x && cat x
/nix/store/2p0c9kjb8wk6brhj5n647ffki2n02n5z-tree-1.8.0
$ nix-env --installed --no-name --out-path --query htop             
error: selector 'htop' matches no derivations
$ which htop
/home/rollf/.nix-profile/bin/htop
$ ll $(which htop)
lrwxrwxrwx 1 rollf users 70 Jan  1  1970 /home/rollf/.nix-profile/bin/htop -> /nix/store/78s3h27xj5bzqib6bwsgss06fc5f2wb3-home-manager-path/bin/htop
```

Do you have any suggestions for this?

**Comment by rycee** [comment]
You can list packages installed by HM by running `home-manager packages`. It is unlikely that a better solution will come in the near future.

On 13 February 2020 14:19:52 CET, "Rolf Schröder" <notifications@github.com> wrote:
>For me, ``nix-env -qaP`` works fine. But when I run ``nix-env -q`` I do
>not get any of the packages I installed through home-manager. I am
>running openSUSE. ``nix-env -q`` only lists packages I installed by
>hand via ``nix-env`` itself.
>
>This is problematic because I cannot query the nix store. For example,
>I have installed ``htop`` via home-manager and ``tree`` with
>``nix-env``:
>
>
>```
>$ nix-env --installed --no-name --out-path --query tree > x && cat x
>/nix/store/2p0c9kjb8wk6brhj5n647ffki2n02n5z-tree-1.8.0
>$ nix-env --installed --no-name --out-path --query htop             
>error: selector 'htop' matches no derivations
>$ which htop
>/home/rollf/.nix-profile/bin/htop
>$ ll $(which htop)
>lrwxrwxrwx 1 rollf users 70 Jan  1  1970
>/home/rollf/.nix-profile/bin/htop ->
>/nix/store/78s3h27xj5bzqib6bwsgss06fc5f2wb3-home-manager-path/bin/htop
>```
>
>Do you have any suggestions for this?
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/rycee/home-manager/issues/266#issuecomment-585750343

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by rollf** [comment]
Hi @rycee , thanks for the quick reply. I stumbled over this because as a resolution of an issue I had, [someone suggested](https://unix.stackexchange.com/a/243189/67833) to run  ``export LOCALE_ARCHIVE="$(nix-env --installed --no-name --out-path --query glibc-locales)/lib/locale/locale-archive"`` where ``glibc-locales`` is a package installed in the nix store (i.e. the intention is to specify a path in the nix-store given only the package name). This fix does not work if I install glibc-locales via home-manager.

**Comment by rycee** [comment]
@rollf If you have `home.package = [ pkgs.glibcLocales ]` in your configuration, wouldn't it then be possible to simply have
```sh
export LOCALE_ARCHIVE="$HOME/.nix-profile/lib/locale/locale-archive"
```
in your shell configuration?

**Comment by rollf** [comment]
@rycee yes, I guess so; I wasn't aware of this. Your feedback is much appreciated!

**Comment by rollf** [comment]
@rycee yes, I guess so; I wasn't aware of this. Your feedback is much appreciated!
